connmgr/connmanager.go

```diff
@@ -530,8 +525,8 @@ func (cm *ConnManager) Connect(ctx context.Context, c *ConnReq) {
 		return
 	}
 
-	if atomic.LoadUint64(&c.id) == 0 {
-		atomic.StoreUint64(&c.id, atomic.AddUint64(&cm.connReqCount, 1))
+	if c.id.Load() == 0 {
+		c.id.Store(cm.connReqCount.Add(1))
```

Comment:
Although you cannot make the `cm.connReqCount` change atomic with `c.id` update, I figure you should still do CAS for `c.id`:

```go
if c.id.CompareAndSwap(0, cm.connReqCount.Add(1)) {
```

Reply:
Ah yes it will evaluate all the input args first.  NVM

Reply:
Mentioned in another comment, but replicating it here for visibility.

I don't think this is really an issue after looking at it more closely.  It would most definitely be an issue if the ID were be assigned anywhere else concurrently, but this is the only place an **existing** connection request has an ID assigned and so hitting a race here would require calling `Connect` on the same connection request concurrently, which doesn't happen.

The only other case of an ID being assigned is in `newConnReq`, but as the name indicates, that creates a **new** connection request and assigns the ID at initialization time, so there is no possible race with this.

Nevertheless, I added an additional mutex in #3055 to prevent potential logic races that could inadvertently be introduced in future code updates by violating that rather hidden and implicit assumption.